### DISCUSSION TITLE
ExceptionHandler은 어떻게 작동하나요?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#162] [Spring MVC]
https://parkmuhyeun.github.io/woowacourse/2023-04-19-Exception-Handler/

**Comment by 3Juhwan** [comment]
`@ExceptionHandler` 애너테이션은 Spring MVC에서 컨트롤러(`@Controller`)나 전역 예외 처리를 위한 `@ControllerAdvice` 클래스의 메서드에서 발생하는 예외를 처리하는 데 사용됩니다. 이 애너테이션은 특정 예외를 처리하는 메서드를 지정하거나 메서드의 파라미터로 처리할 예외를 설정할 수 있습니다.

Spring MVC 애플리케이션에서 예외가 발생하면, DispatcherServlet이 적절한 HandlerExceptionResolver를 찾아 예외를 처리합니다. Spring에 기본적으로 등록된 HandlerExceptionResolver는 세 가지가 있으며, 각 리졸버는 우선순위에 따라 예외를 처리합니다. 그 중 ExceptionHandlerExceptionResolver가 가장 먼저 동작하며, 발생한 예외가 `@ExceptionHandler`에 등록되어 있는지 확인합니다. 만약 처리할 수 없는 예외라면 다음 리졸버로 넘어갑니다. ExceptionHandlerExceptionResolver의 특징은 예외가 WAS로 던져지지 않고 직접 처리된다는 것입니다.

이렇게 함으로써 예외가 발생했을 때 적절한 방법으로 처리되어 사용자에게 친화적인 에러 메시지를 제공하거나 로깅 등의 추가 작업을 수행할 수 있습니다.

**Comment by alstn113** [comment]
스프링 MVC는 컨트롤러(핸들러) 밖으로 예외가 던져진 경우 예외를 해결하고, 동작을 새로 정의할 수 있는 HandlerExceptionResolver를 제공한다. 줄여서 `ExceptionResolver`라고 한다.

스프링은 API 예외 처리 문제를 해결하기 위해 `@ExceptionHandler`라는 애노테이션을 사용하는 예외 처리 기능을 제공하는데, 이것이 바로 `ExceptionHandlerExceptionResolver`이다. 
스프링은 `ExceptionHandlerExceptionResolver`를 기본으로 제공하고, 기본으로 제공하는 `ExceptionResolver`중에 우선순위도 가장 높다. 실무에서 API 예외 처리는 대부분 이 기능을 사용한다. 

`@ControllerAdvice`나 `@Controller` 내부에 처리하고 싶은 예외를 지정해주면 된다. 
@ControllerAdvice -> @Controller 순, 또는 IllegalArgumentException -> RuntimeException 처럼 자세한 것이 우선 순위를 가진다. 

참고로 지정한 예외 또는 그 예외의 자식 클래스를 모두 잡을 수 있다.

참고: 김영한 MVC-2 9장 API 예외 처리
